File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        def dfs(curr, target, start):
            if target <= 0: 
                if target == 0: res.append(curr[:])
                return
            for i in range(start, len(candidates)):
                curr.append(candidates[i])
                dfs(curr, target - candidates[i], i)
                curr.pop()

        dfs([], target, 0)
        return res
            
        


        

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        backtrack(candidates, target, 0, 0, new ArrayList<>(), res);
        return res;
    }

    private void backtrack(int[] candidates, int target, int start, int sum, List<Integer> curr, List<List<Integer>> res) {
        if (sum == target) {
            res.add(new ArrayList<>(curr));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (sum + candidates[i] > target) break;
            curr.add(candidates[i]);
            backtrack(candidates, target, i, sum + candidates[i], curr, res);
            curr.remove(curr.size() - 1);
        }
    }

}